src/net/data_stream.rs

```diff
@@ -73,216 +107,143 @@ pub struct AdcDacData {
-    pub fn new(buf: &'a mut [u8], subsample_rate: usize) -> Self {
+impl FrameGenerator {
+    fn new(
+        queue: Producer<'static, StreamFrame, FRAME_COUNT>,
+        pool: &'static Pool<[u8; FRAME_SIZE]>,
+    ) -> Self {
         Self {
-            buf,
-            start_id: None,
-            num_blocks: 0,
-            subsample_rate,
-            write_index: 4,
+            queue,
+            pool,
+            current_frame: None,
+            sequence_number: 0,
         }
     }
 
-    /// Add a batch of data to the packet.
-    ///
-    /// # Note
-    /// Serialization occurs as the packet is added.
+    /// Add a batch to the current stream frame.
     ///
     /// # Args
-    /// * `batch` - The batch to add to the packet.
-    pub fn add_batch(&mut self, batch: &AdcDacData) -> Result<(), ()> {
-        // Check that the block is sequential.
-        if let Some(id) = &self.start_id {
-            if batch.block_id != id.wrapping_add(self.num_blocks.into()) {
-                return Err(());
+    /// * `format` - The format of the stream. This must be the same for each execution.
+    /// * `f` - A closure that will be provided the buffer to write batch data into. The buffer will
+    ///   be the size of the `T` template argument.
+    pub fn add<F, const T: usize>(&mut self, format: StreamFormat, f: F)
+    where
+        F: FnMut(&mut [u8]),
+    {
+        let sequence_number = self.sequence_number;
+        self.sequence_number = self.sequence_number.wrapping_add(1);
+
+        if self.current_frame.is_none() {
+            if let Some(buffer) = self.pool.alloc() {
+                self.current_frame.replace(StreamFrame::new(
+                    buffer,
+                    format,
+                    sequence_number,
+                ));
+            } else {
+                return;
             }
-        } else {
-            // Otherwise, this is the first block. Record the strt ID.
-            self.start_id = Some(batch.block_id);
-        }
-
-        // Check that there is space for the block.
-        let block_size_bytes = SAMPLE_BUFFER_SIZE / self.subsample_rate * 4 * 2;
-        if self.buf.len() - self.get_packet_size() < block_size_bytes {
-            return Err(());
         }
 
-        // Copy the samples into the buffer.
-        for device in &[batch.adcs, batch.dacs] {
-            for channel in device {
-                for sample in channel.iter().step_by(self.subsample_rate) {
-                    self.buf[self.write_index..self.write_index + 2]
-                        .copy_from_slice(&sample.to_be_bytes());
-                    self.write_index += 2;
-                }
+        assert!(
+            format == self.current_frame.as_ref().unwrap().format,
+            "Unexpected stream format encountered"
+        );
+
+        self.current_frame.as_mut().unwrap().add_batch::<_, T>(f);
+
+        if self.current_frame.as_ref().unwrap().is_full::<T>() {
```

Comment:
Beyond the clippy lint I suspect this can be simplified further with some `if let` or by merging it with the `as_ref()` above. At least there should be a way to cancel a few of the 4 `unwrap()`s here (L230, L234, L236, L239).